src/libraries/System.Private.CoreLib/src/System/Number.NumberToFloatingPointBits.cs

```diff
@@ -361,34 +1082,56 @@ private static ulong NumberToDoubleFloatingPointBits(ref NumberBuffer number, in
             uint integerDigitsPresent = Math.Min(positiveExponent, totalDigits);
             uint fractionalDigitsPresent = totalDigits - integerDigitsPresent;
 
-            uint fastExponent = (uint)(Math.Abs(number.Scale - integerDigitsPresent - fractionalDigitsPresent));
+            // Above 19 digits, we rely on slow path
+            if (totalDigits <= 19)
+            {
+                byte* src = number.GetDigitsPointer();
 
-            // When the number of significant digits is less than or equal to 15 and the
-            // scale is less than or equal to 22, we can take some shortcuts and just rely
-            // on floating-point arithmetic to compute the correct result. This is
-            // because each floating-point precision values allows us to exactly represent
-            // different whole integers and certain powers of 10, depending on the underlying
-            // formats exact range. Additionally, IEEE operations dictate that the result is
-            // computed to the infinitely precise result and then rounded, which means that
-            // we can rely on it to produce the correct result when both inputs are exact.
+                ulong mantissa = DigitsToUInt64(src, (int)(totalDigits));
 
-            byte* src = number.GetDigitsPointer();
+                int exponent = (int)(number.Scale - integerDigitsPresent - fractionalDigitsPresent);
+                int fastExponent = Math.Abs(exponent);
 
-            if ((totalDigits <= 15) && (fastExponent <= 22))
-            {
-                double result = DigitsToUInt64(src, (int)(totalDigits));
-                double scale = s_Pow10DoubleTable[fastExponent];
+                // When the number of significant digits is less than or equal to MaxMantissaFastPath and the
+                // scale is less than or equal to MaxExponentFastPath, we can take some shortcuts and just rely
+                // on floating-point arithmetic to compute the correct result. This is
+                // because each floating-point precision values allows us to exactly represent
+                // different whole integers and certain powers of 10, depending on the underlying
+                // formats exact range. Additionally, IEEE operations dictate that the result is
+                // computed to the infinitely precise result and then rounded, which means that
+                // we can rely on it to produce the correct result when both inputs are exact.
+                // This is known as Clinger's fast path
 
-                if (fractionalDigitsPresent != 0)
+                if ((mantissa <= info.MaxMantissaFastPath) && (fastExponent <= info.MaxExponentFastPath))
```

Comment:
This can be the same for `half`, `float`, and `double` and won't incur any issues due to double rounding/etc (see also https://dl.acm.org/doi/10.1145/221332.221334). That is we should just preserve the equivalent of `if ((totalDigits <= 15) && (fastExponent <= 22))` path.
- Notably the old code capped the mantissa at `999,999,999,999,999`. I don't remember why we did that rather than just checking against the maximum mantissa.

Reply:
The Clinger fast path relies on the fact that if both x and y are exact, then x * y is the correctly rounded product. Similarly with x/y. So if you have an integer that can be represented exactly as the floating type (x) and a power of ten that can be represented exactly as the floating type (y), then x*y is exact. That's the original Clinger's fast path.

I was unaware of https://dl.acm.org/doi/pdf/10.1145/221332.221334

Let me consider Theorem 3 from this paper.

So if I have two low-precision floats (q digits) and I compute a high-precision product (p for p >= 2*q) and then I round that result down to q digits, it is as good as computing the exact (best) q-digit product.

Is this the theorem you are invoking?

Reply:
Correct. It can basically be summed as provided you are doing what is effectively "a single operation" and you are computing it to at least 2x the target precision, you can safely do so.

So for some `float x` and some `float y`, doing `float result = (float)((double)x + (double)y)` will compute the same result as simply `float result = x + y`.

In the same vein, here we have two values that are known to be exactly representable as `double` and so `x` and `y` are the "infinitely precise inputs" (no rounding considerations). We are doing a single operation (multiplication or division) to give the nearest representable `double`. Since this 64-bits and therefore at least 2x the bit width of the destination format (float or half), it is safe to take this result and round it down. If either input was not exactly representable this would not be possible.

Reply:
I do not think that the theorem of this paper implies that though. Not immediately. I am not arguing with you,  but I think that there is a missing mathematical derivation.

Reply:
That is, I believe that this part follows...

So for some `float x` and some `float y`, doing `float result = (float)((double)x + (double)y)` will compute the same result as simply `float result = x + y`.

That's what the paper says... but the rest is an extension.

Reply:
(To be clear, I am not arguing nor saying that you are wrong.)

Reply:
> I do not think that the theorem of this paper implies that though

Right and it is entirely possible that I made a mistake last time I worked through the logic here. 

The general statement from the paper is:
> If `x` and `y` have `p-bit` significands, and `x + y` is 
computed exactly and then rounded to `q` places, 
a second rounding to `p` places will not change 
the answer if `p < (q- 1)/2`. This is true not 
only for addition, but also for multiplication, 
division, and square root.

For this case, the theorem/paper shows that for any number less than `2^24` we can rely on multiplication/division as a `double` to compute the correct result. This allows us to share a path with `double` where the limitation is that all inputs are exactly representable. The extension in logic requires showing that for exactly representable inputs up to `q-bit` significands are likewise exactly representable.

As with the paper for the listed scenario, the consideration here is where `x * y` as a double produces some result where `d[p-1]` is `0`, `d[p]` is `1`, and `d[q]` is 1 and `d[q+n]` (ends in `1`, no trailing data) is all zeros or `d[q]` is `0` and `d[q+n]` are not all zeros (ends in `0`, has trailing data). These are the important cases because it represents the case where double rounding "could" occur and result in an "off by one" error.

You can then consider that since `x` is only some whole integer, that the delta between values is the same for every "window" with the window changing every power of 2 and that all integers less than 2^q are exactly representable that `x` for any value up to `2^q-5`, the significand must end with `0000` and will have no trailing data. In fact, the only case you hit where the significand of `x` ends with `1` for integers less than `2^q` is above `2^q - 2` (for double this is `2^52` where the delta between whole integers becomes `1`).

Likewise, the same goes for the powers of 10 that represent `y` where all `1e16` through `1e22` are greater than `2^53` and it needs to be shown that they are all end in 0 and have no trailing data. This bit is trivial to check and it holds true.

With the original logic I had here, `x` is further restricted to 15 digits and so will always have the lowest 3 bits be zero with no trailing data. `y` will likewise have the lowest 8-bits (or more) be zero with no trailing data for `1e19` and below. It will have the lowest 4-bits be zero for `1e20`/`1e21` and the lowest bit be zero for `1e22`.

-------------

And here's where I need to find my notes where I worked through the math showing that the two 53-bit mantissas given the restrictions above should never produce a value that results in one of the problematic shapes.

Unfortunately, I've moved houses since then and I'm not quite sure where I put them so I'll potentially have to work through the math again.

I don't mind leaving it off if there are concerns here.

Reply:
Will go ahead and close this ask as it will be fine to leave off given the numbers and I don't have the time to work through the math again atm.